ninja-core/src/main/java/ninja/ReverseRouter.java

```diff
@@ -0,0 +1,421 @@
+/*
+ * Copyright 2016 ninjaframework.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ninja;
+
+import com.google.common.base.Optional;
+import java.io.UnsupportedEncodingException;
+import java.lang.reflect.Method;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import javax.inject.Inject;
+import ninja.ControllerMethods.ControllerMethod;
+import ninja.ReverseRouter.Builder;
+import ninja.utils.LambdaRoute;
+import ninja.utils.MethodReference;
+import ninja.utils.NinjaProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Reverse routing. Lookup the uri associated with a controller method.
+ * 
+ * @author Joe Lauer (jjlauer)
+ */
+public class ReverseRouter implements WithControllerMethod<Builder> {
+    static private final Logger log = LoggerFactory.getLogger(ReverseRouter.class);
+    
+    static public class Builder {
+        
+        private final String contextPath;
+        private final Route route;
+        private Map<String,String> paths;
+        private Map<String,String> queries;
+        
+        public Builder(String contextPath, Route route) {
+            this.contextPath = contextPath;
+            this.route = route;
+        }
+
+        public Route getRoute() {
+            return route;
+        }
+        
+        /**
+         * Add a parameter as a path replacement. Will validate the path parameter
+         * exists. This method will URL encode the values when building the final
+         * result.
+         * @param name The path parameter name
+         * @param value The path parameter value
+         * @return A reference to this builder
+         * @see #rawPath(java.lang.String, java.lang.Object) 
+         */
+        public Builder path(String name, Object value) {
```

Comment:
I debated that as well.   The fact you brought it up means we should change.  I like pathParam and queryParam.